core/src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -268,6 +269,20 @@ public Environment getEnvironment() {
      * Start the node. If the node is already started, this method is no-op.
      */
     public Node start() {
+        return start((s, b) -> {});
+    }
+
+    /**
+     * Start the node but invoke the registered callback after network
+     * services are started but before the cluster service is started
+     * and before the network service starts accepting incoming network
+     * requests. If the node is already started, this method is a
+     * no-op.
+     *
+     * @param nodeStartCallback the registered callback
+     * @return the started node
+     */
+    public Node start(BiConsumer<Settings, BoundTransportAddress> nodeStartCallback) {
```

Comment:
I'm not sure about this call back.. feels hacky to pass in such a callback which is fairly obscure as to what it needs to do.

Maye have a method called `validateNodeBeforeAcceptingRequests` which does nothing by default. The bootstrap class can instantiate a new `ValidatedNode` class inheriting from Node (ala what we do in the tests - see `MockNode`) and that class will override the above method to do the checks?